src/string_impls/to_string.rs

```diff
@@ -13,16 +13,358 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use super::formatter::Formatter;
-use super::{string_helpers, FormatOpts};
+use super::string_helpers;
 use crate::prelude::*;
 use crate::value_type::{BitSequence, Composite, Primitive, Value, ValueDef, Variant};
 use core::fmt::{Display, Write};
 
+/// A struct which will try to format a [`Value`] and write the output to a writer.
+/// This can be configured with custom parsers to extend how to write out the value.
+pub struct ToWriterBuilder<T, W> {
+    style: FormatStyle,
+    custom_formatters: Vec<CustomFormatter<T, W>>,
+    indent_by: String,
+    leading_indent: String,
+    print_context: Option<ContextPrinter<T, W>>,
+}
+
+type CustomFormatter<T, W> = Box<dyn Fn(&Value<T>, &mut W) -> Option<core::fmt::Result> + 'static>;
+type ContextPrinter<T, W> = Box<dyn Fn(&T, &mut W) -> core::fmt::Result + 'static>;
+
+impl<T, W: core::fmt::Write> ToWriterBuilder<T, W> {
+    pub(crate) fn new() -> Self {
+        ToWriterBuilder {
+            style: FormatStyle::Normal,
+            custom_formatters: Vec::new(),
+            indent_by: "  ".to_owned(),
+            leading_indent: String::new(),
+            print_context: None,
+        }
+    }
+
+    /// Write the [`crate::Value`] to a compact string, omitting spaces.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use scale_value::{value,Value};
+    /// use scale_value::stringify::to_writer_custom;
+    ///
+    /// let val = value!({
+    ///     foo: (1,2,3,4,5),
+    ///     bar: true
+    /// });
+    ///
+    /// let mut s = String::new();
+    ///
+    /// to_writer_custom()
+    ///     .compact()
+    ///     .write(&val, &mut s)
+    ///     .unwrap();
+    ///
+    /// assert_eq!(s, r#"{foo:(1,2,3,4,5),bar:true}"#);
+    /// ```
+    pub fn compact(mut self) -> Self {
+        self.style = FormatStyle::Compact;
+        self
+    }
+
+    /// Write the [`crate::Value`] to a pretty spaced string.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use scale_value::{value, Value};
+    /// use scale_value::stringify::to_writer_custom;
+    ///
+    /// let val = value!({
+    ///     foo: (1,2,3,4,5),
+    ///     bar: true
+    /// });
+    ///
+    /// let mut s = String::new();
+    ///
+    /// to_writer_custom()
+    ///     .spaced()
+    ///     .write(&val, &mut s)
+    ///     .unwrap();
+    ///
+    /// assert_eq!(s, r#"{
+    ///   foo: (
+    ///     1,
+    ///     2,
+    ///     3,
+    ///     4,
+    ///     5
+    ///   ),
+    ///   bar: true
+    /// }"#);
+    /// ```
+    pub fn spaced(mut self) -> Self {
```

Comment:
nit: indented

Reply:
Mmm spaced isn't the best word for this is it! I was also thinking `pretty` to align with eg serde_json's use of the word!

Reply:
Changed to `pretty`; thanks for pointing this out!